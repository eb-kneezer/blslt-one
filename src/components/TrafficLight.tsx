import styled, { css, keyframes } from "styled-components";

type TrafficLightPosition = "top" | "right" | "bottom" | "left";
type TrafficLightType = "master" | "slave";

type MainStateColors = "red" | "green";
export interface ControlState {
  prev: MainStateColors;
  curr: MainStateColors;
}

interface TrafficLightComponentProps {
  position: TrafficLightPosition;
  type: TrafficLightType;
  isTransiting: boolean;
  state: ControlState;
}

export const TrafficLightComponent: React.FC<TrafficLightComponentProps> = ({
  position,
  type,
  isTransiting,
  state,
}) => {
  const isRed =
    type === "master" ? state.curr === "red" : state.curr === "green";
  const isGreen =
    type === "master" ? state.curr === "green" : state.curr === "red";

  const style: React.CSSProperties = {
    top: position === "top" ? "20px" : position === "bottom" ? "auto" : "50%",
    bottom: position === "bottom" ? "20px" : "auto",
    left: position === "left" ? "20px" : position === "right" ? "auto" : "50%",
    right: position === "right" ? "20px" : "auto",
    transform: `${
      position === "left" || position === "right"
        ? "translateY(-50%)"
        : "translateX(-50%)"
    }`,
  };

  const indicatorStyle: React.CSSProperties = {
    top: position === "top" ? "100%" : "auto",
    bottom: position === "bottom" ? "100%" : "auto",
    left: position === "left" ? "100%" : "auto",
    right: position === "right" ? "100%" : "auto",
  };

  return (
    <TrafficLightWrapper data-testid={`traffic-light-${type}`} style={style}>
      <TrafficLight>
        <Light
          data-testid="red-light"
          $color="red"
          $active={isRed && !isTransiting}
        />
        <Light
          data-testid="yellow-light"
          $color="yellow"
          $active={isTransiting}
        />
        <Light
          data-testid="green-light"
          $color="green"
          $active={isGreen && !isTransiting}
        />
        <PedestrianIndicator
          style={indicatorStyle}
          $state={isRed ? "walk" : "stop"}
          $transiting={isTransiting}
        >
          {isRed ? "ðŸ‘Ÿ" : "âœ‹"}
        </PedestrianIndicator>
      </TrafficLight>
    </TrafficLightWrapper>
  );
};

const TrafficLightWrapper = styled.div`
  position: absolute;
`;

const TrafficLight = styled.div`
  width: 40px;
  background-color: #fbc336;
  display: flex;
  flex-direction: column;
  gap: 5px;
  justify-content: space-around;
  align-items: center;
  padding: 5px 0;
  border-radius: 8px;
  position: relative;
`;

interface LightProps {
  $color: string;
  $active: boolean;
}

const Light = styled.div<LightProps>`
  width: 30px;
  height: 30px;
  border-radius: 100%;
  background-color: ${(props) => (props.$active ? props.$color : "#c89c2c")};
`;

interface PedestrianIndicatorProps {
  $state: "stop" | "walk";
  $transiting: boolean;
}

const blink = keyframes`
  0%, 100% {
    opacity: 1;
  }
  50% {
    opacity: 0;
  }
`;

const PedestrianIndicator = styled.div<PedestrianIndicatorProps>`
  font-size: 25px;
  position: absolute;
  ${(props) =>
    props.$transiting &&
    css`
      animation: ${blink} 1s ease-in-out infinite;
    `}
`;
