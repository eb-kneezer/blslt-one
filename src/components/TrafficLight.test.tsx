import { render } from "@testing-library/react";
import { TrafficLightComponent } from "./TrafficLight";

describe("TrafficLight", () => {
  // Renders correctly with default props
  it("should render correctly with default props", () => {
    const { container } = render(
      <TrafficLightComponent
        position="top"
        type="master"
        isTransiting={false}
        state={{ curr: "red", prev: "green" }}
      />
    );
    expect(container).toMatchSnapshot();
  });

  // Display red light when curr state is red and type is master
  it("should display red light when state is red and type is master", () => {
    const { container } = render(
      <TrafficLightComponent
        position="top"
        type="master"
        isTransiting={false}
        state={{ curr: "red", prev: "green" }}
      />
    );
    const redLight = container.querySelector('[data-testid="red-light"]');
    expect(redLight).toHaveStyle("background-color: rgb(255, 0, 0)");
  });

  // Displays green light when curr state is green and type is master
  it("should display green light when state is green and type is master", () => {
    const { container } = render(
      <TrafficLightComponent
        position="top"
        type="master"
        isTransiting={false}
        state={{ curr: "green", prev: "red" }}
      />
    );
    const greenLight = container.querySelector('[data-testid="green-light"]');
    expect(greenLight).toHaveStyle("background-color: rgb(0, 128, 0)");
  });

  // Display red light when curr state is red and type is slave
  it("should display green light when state is red and type is slave", () => {
    const { container } = render(
      <TrafficLightComponent
        position="top"
        type="slave"
        isTransiting={false}
        state={{ curr: "red", prev: "green" }}
      />
    );
    const greenLight = container.querySelector('[data-testid="green-light"]');
    expect(greenLight).toHaveStyle("background-color: rgb(0, 128, 0)");
  });

  // Displays green light when curr state is green and type is slave
  it("should display red light when state is green and type is slave", () => {
    const { container } = render(
      <TrafficLightComponent
        position="top"
        type="slave"
        isTransiting={false}
        state={{ curr: "green", prev: "red" }}
      />
    );
    const redLight = container.querySelector('[data-testid="red-light"]');
    expect(redLight).toHaveStyle("background-color: rgb(255, 0, 0)");
  });

  // Shows yellow light when isTransiting is true
  it("should show yellow light when isTransiting is true", () => {
    const { container } = render(
      <TrafficLightComponent
        position="top"
        type="master"
        isTransiting={true}
        state={{ curr: "green", prev: "red" }}
      />
    );
    const yellowLight = container.querySelector('[data-testid="yellow-light"]');
    expect(yellowLight).toHaveStyle("background-color: rgb(255, 255, 0)");
  });

  // Pedestrian indicator shows walk when light is red
  it("should show walk indicator when light is red", () => {
    const { getByText } = render(
      <TrafficLightComponent
        position="top"
        type="master"
        isTransiting={false}
        state={{ curr: "red", prev: "green" }}
      />
    );
    const walkIndicator = getByText("ðŸ‘Ÿ");
    expect(walkIndicator).toBeInTheDocument();
  });

  // Pedestrian indicator shows stop when light is green
  it("should show stop pedestrian indicator when light is green", () => {
    const { getByText } = render(
      <TrafficLightComponent
        position="top"
        type="master"
        isTransiting={false}
        state={{ curr: "green", prev: "red" }}
      />
    );
    expect(getByText("âœ‹")).toBeInTheDocument();
  });
});
