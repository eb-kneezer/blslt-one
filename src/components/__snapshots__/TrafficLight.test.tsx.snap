// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TrafficLight > should render correctly with default props 1`] = `
<div>
  <div
    class="sc-blHHSb kfSbpr"
    data-testid="traffic-light-master"
    style="top: 20px; left: 50%; transform: translateX(-50%);"
  >
    <div
      class="sc-gtLWhw FreFW"
    >
      <div
        class="sc-egkSDF cZsofJ"
        data-testid="red-light"
      />
      <div
        class="sc-egkSDF dlBHlB"
        data-testid="yellow-light"
      />
      <div
        class="sc-egkSDF dlBHlB"
        data-testid="green-light"
      />
      <div
        class="sc-fAUdSK cqjXRm"
        style="top: 100%;"
      >
        ðŸ‘Ÿ
      </div>
    </div>
  </div>
</div>
`;
